### PR TITLE
Improve code and rewrite Token to separate classes with self-validation

### PR DESCRIPTION
**Pull request by ntdesmond**

### CODE REVIEW

File: O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿using OCompiler.Extensions;
+
+namespace OCompiler.Analyze.Lexical.Tokens
+{
+    abstract class Token
+    {
+        public long StartOffset { get; }
+        public int Length { get; }
```
Comment (DOCtorActoAntohich) [inline O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: Rewrite getter to access `Literal.Length` instead

File: O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            Literals.Boolean   => new BooleanLiteral(position, literal.Value),
+            Literals.Delimiter => new Delimiter(position, literal.Value),
+            Literals.Keyword   => new Keyword(position, literal.Value),
+            _ => throw new System.ArgumentException("Argument passed is not a reserved literal")
+        };
+
+        public static bool TryParse(long position, string literal, out Token token)
+        {
+            var reservedLiteral = Literals.ReservedLiteral.GetByValue(literal);
+            if (reservedLiteral == Literals.ReservedLiteral.Empty)
```
Comment (DOCtorActoAntohich) [inline O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: Is it possible to make code less nested at the cost of more complex boolean expressions?

File: O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                if (literal[0].IsIdentifierOrNumber())
+                {
+                    if (char.IsDigit(literal[0]) && int.TryParse(literal, out int _))
+                    {
+                        token = new IntegerLiteral(position, literal);
+                        return true;
+                    }
+                    else if (literal.Contains('.'))
+                    {
+                        if (literal.ToDouble(out double _))
```
Comment (DOCtorActoAntohich) [inline O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: What if this one is false?

File: O-Compiler/Source/Analyze/Lexical/Tokens/RealLiteral.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    internal class RealLiteral : Token
+    {
+        public double Value { get; }
+        public RealLiteral(long startOffset, string literal) : base(startOffset, literal)
+        {
+            if (literal.ToDouble(out double result))
+            {
+                Value = result;
+            }
+            else
```
Comment (DOCtorActoAntohich) [inline O-Compiler/Source/Analyze/Lexical/Tokens/RealLiteral.cs]: Avoid using `else`

File: O-Compiler/Source/Analyze/Lexical/Tokens/IntegerLiteral.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public int Value { get; }
+        public IntegerLiteral(long startOffset, string literal) : base(startOffset, literal)
+        {
+            if (literal.ToInteger(out int result))
+            {
+                Value = result;
+            }
+            else
+            {
```
Comment (DOCtorActoAntohich) [inline O-Compiler/Source/Analyze/Lexical/Tokens/IntegerLiteral.cs:11]: Avoid using `else`

File: O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs (hunk 3 above)
Comment (ntdesmond) [reply O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: > What if this one is false?

no return => all the way to bottom and return `false`

File: O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs (hunk 2 above)
Comment (ntdesmond) [reply O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: ~~at the cost of ternaries~~
will see

Comment (ntdesmond) [reply O-Compiler/Source/Analyze/Lexical/Tokens/Token.cs]: Fixed with https://github.com/DOCtorActoAntohich/O-Compiler/pull/4/commits/ce64e8110617c32e2d17b69a0c637a7e74da7a3a

**Review by DOCtorActoAntohich — APPROVED** [review]
Naisu

**Pull request merged**